packages/@aws-cdk/aws-synthetics/lib/code.ts

```diff
@@ -0,0 +1,177 @@
+   * @param bucket The S3 bucket
+   * @param key The object key
+   * @param objectVersion Optional S3 object version
+   */
+  public static fromBucket(bucket: s3.IBucket, key: string, objectVersion?: string): S3Code {
+    return new S3Code(bucket, key, objectVersion);
+  }
+
+  /**
+   * Called when the canary is initialized to allow this object to bind
+   * to the stack, add resources and have fun.
+   *
+   * @param scope The binding scope. Don't be smart about trying to down-cast or
+   * assume it's initialized. You may just use it as a construct scope.
+   */
+  public abstract bind(scope: Construct, handler: string): CodeConfig;
+}
+
+/**
+ * Configuration of the code class
+ */
+export interface CodeConfig {
+  /**
+   * The location of the code in S3 (mutually exclusive with `inlineCode`).
+   *
+   * @default - mutually exclusive with `inlineCode`
+   */
+  readonly s3Location?: s3.Location;
+
+  /**
+   * Inline code (mutually exclusive with `s3Location`).
+   *
+   * @default - mutually exclusive with `s3Location`
+   */
+  readonly inlineCode?: string;
+}
+
+/**
+ * Canary code from an Asset
+ */
+export class AssetCode extends Code {
+  private asset?: s3_assets.Asset;
+
+  /**
+   * @param assetPath The path to the asset file or directory.
+   */
+  constructor(private assetPath: string, private options?: s3_assets.AssetOptions){
+    super();
+  }
+
+  public bind(scope: Construct, handler: string): CodeConfig {
+    this.validateCanaryAsset(handler);
+
+    // If the same AssetCode is used multiple times, retain only the first instantiation.
+    if (!this.asset){
+      this.asset = new s3_assets.Asset(scope, 'Code', {
+        path: this.assetPath,
+        ...this.options,
+      });
+    }
+
+    return {
+      s3Location: {
+        bucketName: this.asset.s3BucketName,
+        objectKey: this.asset.s3ObjectKey,
+      },
+    };
+  }
+
+  /**
+   * Validates requirements specified by the canary resource. For example, the canary code with handler `index.handler`
+   * must be found in the file structure `nodejs/node_modules/index.js`.
+   *
+   * Requires path to be either zip file or directory.
+   * Requires asset directory to have the structure 'nodejs/node_modules'.
+   * Requires canary file to be directly inside node_modules folder.
+   * Requires canary file name matches the handler name.
+   * @see https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/CloudWatch_Synthetics_Canaries_WritingCanary.html
+   *
+   * @param handler the canary handler
+   */
+  private validateCanaryAsset(handler: string){
+    if(path.extname(this.assetPath) !== '.zip'){
+      if(path.extname(this.assetPath) !== ''){
+        throw new Error(`Asset must be a .zip file or a directory (${this.assetPath})`);
+      }
+      const filename = handler.substring(0,handler.indexOf('.')) + '.js';
+      if(!fs.existsSync(path.join(this.assetPath,'nodejs', 'node_modules', filename))){
```

Comment:
The canary requires a folder structure like `nodejs/node_modules/canary.js`. Without the structure, the canary successfully deploys but results in a FAILED state at _runtime_ -- potentially triggering alarms and all that good stuff. Obviously, that is awful, so the error checking here is an okay solution... for now.

This checks to see if `nodejs/node_modules/file.js` is inside of the directory given.

However, an even better solution may be to allow the user to specify a single file and then build the `nodejs/node_modules` structure in front of it before bundling in the asset. This would require playing around with the `asset` module and likely adding a prop there that triggers the necessary structure appended to the file name. 

Users would then specify something like:

`synthetics.Code.fromAsset(file.js)`

and we would append the necessary structure to make this work.